datashader/core.py

```diff
@@ -348,11 +350,12 @@ def raster(self,
             data = data.filled()
 
         # Restore original dtype
-        data = data.astype(dtype)
+        if dtype != data.dtype:
+            data = data.astype(dtype)
 
         # Compute DataArray metadata
         xs, ys = compute_coords(self.plot_width, self.plot_height, self.x_range, self.y_range, res)
-        coords = {xdim: xs.astype(xvals.dtype), ydim: ys.astype(yvals.dtype)}
+        coords = {xdim: xs, ydim: ys}
```

Comment:
I added this before but realized it will result in poor behavior when downsampling or upsampling with integer coordinates. The output coordinates will just always be floats.